File: sum_path.cpp
//TC: O(n) where n is number of nodes
//SC: O(n) where n is number of nodes.

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    vector<vector<int>> paths;
    //pass vector by reference. 
    void findPaths(TreeNode* node, int target, int curr, vector<int> &path) {
        if (node == NULL)
            return;
        curr += node->val;
        path.push_back(node->val);
        if (curr == target && node->left == node->right)
            paths.push_back(path);
        
        findPaths(node->left, target, curr, path);
        findPaths(node->right, target, curr, path);
        //when we reach a left leaf node, pop the leaf node to go back to root.
        path.pop_back();
    }
    
    
    
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int> path;
        findPaths(root, targetSum, 0, path);
        return paths;
    }
};

File: SumPath.py
#time complexity: O(n)
#Space complexity: O(1)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    res=[]
    count=0
    def pathSum(self, root, targetSum: int):
        self.res=[]
        self.helper(root,[],0,targetSum)
        return self.res
    def helper(self,root,path,currentSum,targetSum):
        if root== None:
            return
        path.append(root.val)
        if root.left==None and root.right==None and currentSum+root.val==targetSum:
            self.res.append(path.copy())
        self.helper(root.left,path,currentSum + root.val,targetSum)
        self.helper(root.right,path,currentSum + root.val,targetSum)
        path.pop()